### PR TITLE
feat(pubsub): flow control for Publisher

### LINKED ISSUE #5607: Implement flow control for Publisher::Publish()
We want to help applications flow control their publication of messages to avoid problems such as:

- Excessive memory consumed by pending `AsyncPublish()` batches.
- Duplicate messages due to retransmission of `AsyncPublish()` requests that receive a (false) deadline exceeded because the publisher is too busy.

The application should be able to specify:

- The maximum number of messages pending in the publisher.
- The maximum number of pending bytes (the sum of the size of all pending messages) in the publisher.
- The action to take when these limits are exceeded.

Possible actions to take include:

- Ignore the problem, just proceed as before, this is equivalent to setting the limits to "infinity".
- Block the caller until there is room to publish more messages.
- Return a future satisfied with an error for the message(s) that would exceed these limits.

The options to control these limits should be set when the publisher is created. The names of these options should match the names for the subscriber flow control options.

### PR DESCRIPTION
**Pull request by coryan**
Some applications may want to constrain the messages "in flight" sent
from a publisher. This adds configuration parameters to set such
constraints, and either block until the messages are received by the
service, or reject messages that would exceed the constraints.

Fixes #5607

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6678)

### CODE REVIEW

**Review by devbww — APPROVED** [review]

File: google/cloud/pubsub/publisher_options.h [hunk 1]
```diff
@@ -132,16 +155,84 @@ class PublisherOptions {
...
 
  private:
   static auto constexpr kDefaultMaximumHoldTime = std::chrono::milliseconds(10);
   static std::size_t constexpr kDefaultMaximumMessageCount = 100;
   static std::size_t constexpr kDefaultMaximumMessageSize = 1024 * 1024L;
+  static std::size_t constexpr kDefaultMaximumPendingBytes =
+      (std::numeric_limits<std::size_t>::max)();
+  static std::size_t constexpr kDefaultMaximumPendingMessages =
+      (std::numeric_limits<std::size_t>::max)();
+  // See https://cloud.google.com/pubsub/pricing, the minimum legal size for a
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Fixed.

File: google/cloud/pubsub/publisher_options.h [hunk 2]
```diff
@@ -132,16 +155,84 @@ class PublisherOptions {
...
+  /**
+   * @name Publisher flow control.
+   *
+   * After a publisher flushes a batch of messages the batch is (obviously) not
+   * received immediately by the service. While the batch remains pending it
+   * potentially consumes memory resources in the client (and/or the service).
+   *
+   * Some applications may have constraints on the number of bytes and/or
+   * messages they can tolerate in this pending state, and may prefer to block
+   * or reject messages
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Fixed.

File: google/cloud/pubsub/publisher_options.h [hunk 3]
```diff
@@ -109,6 +117,21 @@ class PublisherOptions {
...
+  //@}
+
+  //@{
+  /**
+   * @name Publisher message ordering.
+   *
+   * To guarantee messages are received by the service in the same order that
+   * the application gives them to a publisher the client library needs to wait
+   * until a batch of messages is successfully delivered before sending the next
+   * batch. Otherwise batches may arrive out of order as the is no guarantee the
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Fixed.

File: google/cloud/pubsub/publisher_options.h [hunk 4]
```diff
@@ -109,6 +117,21 @@ class PublisherOptions {
...
     return *this;
   }
+  //@}
+
+  //@{
+  /**
+   * @name Publisher message ordering.
+   *
+   * To guarantee messages are received by the service in the same order that
+   * the application gives them to a publisher the client library needs to wait
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Fixed

File: google/cloud/pubsub/publisher_options.h [hunk 5]
```diff
@@ -47,7 +48,14 @@ class PublisherOptions {
...
   PublisherOptions() = default;
 
-  /// The maximum hold time.
+  //@{
+  /**
+   * @name Publisher batch control
+   *
+   * It is more efficient (in terms of client CPU and client network usage) to
+   * send multiple messages in a single "batch" to the service. The following
+   * configuration options can be used To improve throughput (as
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Fixed.

File: google/cloud/pubsub/internal/flow_controlled_publisher_connection.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  void OnPublish(std::size_t message_size);
+  bool IsFull() const {
+    return pending_bytes_ >= options_.maximum_pending_bytes() ||
+           pending_messages_ >= options_.maximum_pending_messages();
+  }
+  bool MakesFull(std::size_t message_size) const {
+    if (pending_messages_ + 1 > options_.maximum_pending_messages()) {
+      return true;
+    }
+    return pending_bytes_ + message_size > options_.maximum_pending_bytes();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/flow_controlled_publisher_connection.h]: Fixed, in the process I noticed that `IsFull()` was using `>=` and `MakesFull()` was using `>`. The latter is correct, fixed.

File: google/cloud/pubsub/internal/flow_controlled_publisher_connection.h [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  void ResumePublish(ResumePublishParams p) override;
+
+  std::size_t max_pending_messages() const {
+    std::unique_lock<std::mutex> lk(mu_);
+    return max_pending_messages_;
+  }
+  std::size_t max_pending_bytes() const {
+    std::unique_lock<std::mutex> lk(mu_);
+    return max_pending_bytes_;
+  }
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/flow_controlled_publisher_connection.h:58]: Yeah, they are only intended for testing.  Added a comment.

File: google/cloud/pubsub/internal/flow_controlled_publisher_connection_test.cc [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+                          .full_publisher_blocks()
+                          .set_maximum_pending_messages(kExpectedMaxMessages));
+  EXPECT_LE(actual->max_pending_messages(), kExpectedMaxMessages);
+}
+
+}  // namespace
+}  // namespace GOOGLE_CLOUD_CPP_PUBSUB_NS
+}  // namespace pubsub_internal
+}  // namespace cloud
+}  // namespace google
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/flow_controlled_publisher_connection_test.cc]: Fixed.

File: google/cloud/pubsub/internal/flow_controlled_publisher_connection.cc [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+namespace {
+StatusOr<std::string> RejectMessage() {
+  return Status(StatusCode::kFailedPrecondition, "Publisher is full");
+}
+}  // namespace
+
+future<StatusOr<std::string>> FlowControlledPublisherConnection::Publish(
+    PublishParams p) {
+  std::unique_lock<std::mutex> lk(mu_);
+  auto const message_size = MessageSize(p.message);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/flow_controlled_publisher_connection.cc:30]: Doh! I started like that and somehow messed it up.

File: google/cloud/pubsub/publisher_options.h [hunk 10]
```diff
@@ -132,16 +155,84 @@ class PublisherOptions {
...
+  std::size_t maximum_pending_messages() const {
+    return maximum_pending_messages_;
+  }
+
+  static int constexpr kFullPublisherIgnored = 0;
+  static int constexpr kFullPublisherBlocks = 1;
+  static int constexpr kFullPublisherRejects = 2;
+
+  /// The current action for a full publisher
+  int full_publisher_action() const { return full_publisher_action_; }
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Not really, changed to an enum, but avoid exposing it as part of the API.

File: google/cloud/pubsub/publisher_options.h [hunk 11]
```diff
@@ -54,7 +54,11 @@ class PublisherOptions {
    *
    * It is more efficient (in terms of client CPU and client network usage) to
    * send multiple messages in a single "batch" to the service. The following
-   * configuration options can be used To improve throughput (as
+   * configuration options can be used to improve throughput: sending larger
+   * batches reduces CPU and network overhead. Note that batches are subject
+   * to [quota limits]
```
Comment (devbww) [inline google/cloud/pubsub/publisher_options.h:59]: s/$/./

File: google/cloud/pubsub/internal/flow_controlled_publisher_connection.h [hunk 12]
```diff
@@ -59,23 +65,15 @@ class FlowControlledPublisherConnection
...
-  }
-  bool RejectWhenFull() const {
-    return options_.full_publisher_action() ==
-           pubsub::PublisherOptions::kFullPublisherRejects;
-  }
-  bool BlockWhenFull() const {
-    return options_.full_publisher_action() ==
-           pubsub::PublisherOptions::kFullPublisherBlocks;
+    return pending_messages_ + 1 > options_.maximum_pending_messages() ||
+           pending_bytes_ + message_size > options_.maximum_pending_bytes();
```
Comment (devbww) [inline google/cloud/pubsub/internal/flow_controlled_publisher_connection.h]: Q. Perhaps the first message should never count as "MakesFull", no matter what its size?  Might that alleviate the "leads to deadlock" issue?

**Review by devbww — APPROVED** [review]

File: google/cloud/pubsub/internal/flow_controlled_publisher_connection.h (hunk 12 above)
Comment (coryan) [reply google/cloud/pubsub/internal/flow_controlled_publisher_connection.h]: Good idea, done.

File: google/cloud/pubsub/publisher_options.h (hunk 11 above)
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Fixed.

**Review by devbww — APPROVED** [review]

**Pull request merged**